File: 0505-The Maze II/0505-The Maze II.cpp
class Solution {
public:
    int shortestDistance(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        int m = maze.size();
        int n = maze[0].size();
        priority_queue<Cell, vector<Cell>, cmp> Q;
        vector<vector<int>> distance(m, vector<int>(n, INT_MAX));
        Q.emplace(start[0], start[1], 0);
        distance[start[0]][start[1]] = 0;
        vector<int> dx = {-1, 1, 0, 0};
        vector<int> dy = {0, 0, -1, 1};
        while (!Q.empty())
        {
            int x = Q.top().row;
            int y = Q.top().col;
            int d = Q.top().dist;
            Q.pop();
            if (x == destination[0] && y == destination[1])
            {
                return d;
            }
            
            for (int i = 0; i < dx.size(); ++i)
            {
                int nx = x + dx[i];
                int ny = y + dy[i];
                int nd = d + 1;
                while (nx >= 0 && nx < m && ny >= 0 && ny < n && maze[nx][ny] == 0)
                {
                    nx += dx[i];
                    ny += dy[i];
                    ++nd;
                }
                
                nx -= dx[i];
                ny -= dy[i];
                --nd;
                
                if (distance[nx][ny] > nd)
                {
                    Q.emplace(nx, ny, nd);
                    distance[nx][ny] = nd;
                }
            }
        }
        
        return -1;
    }
    
private:
    struct Cell {
        int row;
        int col;
        int dist;
        Cell(int r, int c, int d): row(r), col(c), dist(d){}
    };
    
    struct cmp {
        bool operator() (Cell& c1, Cell& c2) {
            return c1.dist > c2.dist;
        }
    };
};

File: 0501-0600/0505-The Maze II/0505-The Maze II.py
import heapq
import math
# Dijkstra
class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        pq = [[0] + start]
        distance = [[math.inf] * n for _ in range(m)]
        distance[start[0]][start[1]] = 0
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        while pq:
            d, r, c = heapq.heappop(pq)
            if [r, c] == destination:
                return d
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]
                nd = d + 1
                while 0 <= nr < m and 0 <= nc < n and maze[nr][nc] == 0:
                    nr += dr[i]
                    nc += dc[i]
                    nd += 1
                nr -= dr[i]
                nc -= dc[i]
                nd -= 1

                if distance[nr][nc] > nd:
                    heapq.heappush(pq, [nd, nr, nc])
                    distance[nr][nc] = nd
        
        return -1


# A*
class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        pq = [[abs(destination[0] - start[0]) + abs(destination[1] - start[1]), 0] + start]
        distance = [[math.inf] * n for _ in range(m)]
        distance[start[0]][start[1]] = 0
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]
        while pq:
            _, d, r, c = heapq.heappop(pq)
            if [r, c] == destination:
                return d
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]
                nd = d + 1
                while 0 <= nr < m and 0 <= nc < n and maze[nr][nc] == 0:
                    nr += dr[i]
                    nc += dc[i]
                    nd += 1
                nr -= dr[i]
                nc -= dc[i]
                nd -= 1

                if distance[nr][nc] > nd:
                    heapq.heappush(pq, [nd + abs(destination[0] - nr) + abs(destination[1] - nc), nd, nr, nc])
                    distance[nr][nc] = nd
        
        return -1
